src/ht.c

```diff
@@ -1,239 +1,238 @@
+	ut32 hash = ht->hashfn (key);
+	ut32 bucket;
+	bool found = true;
+	if (update) {
+		(void)ht_delete_internal (ht, key, &hash);
+	} else {
+		(void)ht_find (ht, key, &found);
+	}
+	if (update || !found) {
+		kvp = calloc (1, sizeof (SdbKv));
+		if (kvp) {
+			if (ht->dupkey) {
+				kvp->key = ht->dupkey (key);
+			} else {
+				kvp->key = (char*)key;
+			}
+			if (ht->dupvalue) {
+				kvp->value = ht->dupvalue (value);
+			} else {
+				kvp->value = (char*)value;
+			}
+			bucket = hash % ht->size;
+			kvp->expire = 0;
+			if (ht->calcsize) {
+				kvp->value_len = (ut32)ht->calcsize (kvp->value);
+			} else {
+				kvp->value_len = 0;
+			}
+			ls_prepend (ht->table[bucket], kvp);
+			ht->count++;
+			// Check if we need to grow the table.
+			if (ht->count >= ht->load_factor * ht_primes_sizes[ht->prime_idx]) {
+				if (ht->prime_idx < sizeof (ht_primes_sizes)/sizeof (ht_primes_sizes[0])) {
+					ht->prime_idx++;
+					internal_ht_grow (ht);
+				}
+			}
+			return true;
+		}
 	}
+	return false;
 }
 
-void *ht_lookup(SdbHash *ht, ut32 hash) {
-	SdbHashEntry *entry;
-	if (!ht || !ht->list || !ht->list->head) return NULL;
-	entry = ht_search (ht, hash);
-	return entry? entry->data : NULL;
+// Inserts the key value pair key, value into the hashtable.
+// Doesn't allow for "update" of the value.
+bool ht_insert(SdbHash* ht, const char* key, const char* value) {
+	return internal_ht_insert (ht, false, key, value);
 }
 
-#if 0
-void ht_set(SdbHash *ht, ut32 hash, void *data) {
-	SdbHashEntry *e = ht_search (ht, hash);
-	if (e) {
-		if (ht->list->free)
-			ht->list->free (e->data);
-		e->data = data;
-		e->iter->data = data;
-	} else ht_insert (ht, hash, data, NULL);
+// Inserts the key value pair key, value into the hashtable.
+// Does allow for "update" of the value.
+bool ht_update(SdbHash* ht, const char* key, const char* value) {
+	return internal_ht_insert (ht, true, key, value);
 }
-#endif
-
-/**
- * Inserts the data with the given hash into the table.
- *
- * Note that insertion may rearrange the table on a resize or rehash,
- * so previously found hash_entries are no longer valid after this function.
- */
-int ht_insert(SdbHash *ht, ut32 hash, void *data, SdbListIter *iter) {
-	ut32 hash_address;
-	if (!ht || !data)
-		return 0;
 
-	if (ht->entries >= ht->max_entries)
-		ht_rehash (ht, ht->size_index + 1);
-	else if (ht->deleted_entries + ht->entries >= ht->max_entries)
-		ht_rehash (ht, ht->size_index);
-
-	hash_address = hash % ht->size;
-	do {
-		SdbHashEntry *entry = ht->table + hash_address;
-		ut32 double_hash;
-
-		if (!entry_is_present (entry)) {
-			if (entry_is_deleted (entry))
-				ht->deleted_entries--;
-			entry->hash = hash;
-			entry->data = data;
-			entry->iter = rehash? iter: ls_append (ht->list, data);
-			ht->entries++;
-			return 1;
+bool ht_insert_kvp(SdbHash* ht, SdbKv* kvp, bool update) {
+	bool found;
+	ut32 bucket, hash = ht->hashfn (kvp->key);
+	if (update) {
+		(void)ht_delete_internal (ht, kvp->key, &hash);
```

Comment:
because if it is `NULL` then we have to compute the hash ourself, otherwise, we take the value of the `*hash`.